### PR TITLE
[otbn] Fix off-by-one error in otbn-as li handling

### LINKED ISSUE #3698: [otbn] load constant > 2047 and < 4096 to gpr
So I want to load the constant 2272 to a GPR.

When using the pseudo instruction: `li x21, 2272`
I get the following error when building: `Error: illegal operands addi x21,x0,2272`

Apparently this is directly replaced by the `addi` since the immediate is smaller than 2^12. However `addi` expects a signed immediate in the assembly, so 2272 needs to be written as -1824 to get the intended binary representation in the reg.

So it could be that that this just how it is specified in riscv assembly, that you need to provide the immediate as a signed number. But what confuses me is that if I add 4096 to the constant and do:
`li x21, 6368`
this produces:
```
lui	x21,0x2
addi	x21,x21,-1824
```

Here the assembler creates the negative immediate correctly (the `0x2` seems not correct though). So it seems that any positive number, but anything in the range 2048 to 4095, is accepted. This seems odd to me. But maybe there is some way to encode the literal I am not aware of. So not 100% sure whether this is a bug.

I would have thought that this translation is handled by the riscv (not otbn) assembler but since it shows up in the `*.dis` file, maybe I am wrong.

### PR DESCRIPTION
**Pull request by rswarbrick**
The "this is a small constant" check is supposed to check whether the
value can be represented as a 12-bit signed number, but it was
checking against the range `[-2^12, 2^12-1]` rather than `[-2^11, 2^11-1]`.

Fixes #3698.

EDIT: I've added a second patch to the back of this PR, which adds some simple tests for `li` expansion. Mainly because I was worried I'd still got something wrong!

### CODE REVIEW

**Review by felixmiller — CHANGES_REQUESTED** [review]
Thanks @rswarbrick, this LGTM and fixes #3698 for me.

However, I came across another related issue when I was testing and I think it is probably worth fixing it here:

I realized that loading `0xFFFFFFFF` fails and actually anything larger than 0xFFFFF7FFF fails, i.e. anything where the MSB of the lower 12 bit that are handled by the `addi` is set.
So I think here, this should again only be expanded to a single `addi` instruction with the sign extend handling the rest.

**Comment by rswarbrick** [comment]
Huh. It looks like the binutils assembler allows any immediate for `li` in the range `[-2^32, 2^32-1]`, wrapping both. So these pairs of lines are equivalent:
```
  li x3, 0xffffffff
  li x3, -1

  li x4, -0x80000001
  li x4, 0x7fffffff
```
I think this is a bit bonkers (especially the negative wrapping). And it gets worse! `li x5, -0x100000001` (the constant is `-(2^32+1)`) assembles as `lui x5, 0; addi x5, x5, -1` (with a pointless `lui`)

I'll file a follow-up patch to allow anything in the range `[-2^31, 2^32-1]`, which allows negative values and big hex constants. I can't really see a use for anything else, and I'm not convinced that binutils is being particularly principled about it!

**Review by felixmiller — APPROVED** [review]
Thanks, this LGTM and the range makes sense from the otbn sw perspective.

So far I cannot think of a situation where we want to load a negative constant. The constants we load are either pointers (into regfile or dmem) or part of a cryptographic constant. For the latter they are always only a share of a WLEN sized number and since we do not do 256 bit sign extend we always handle them as unsigned binary.
Since we are supporting offsets there could be the slight chance that we want a negative relative address for loading from dmem in the future though.

File: hw/ip/otbn/dv/otbnsim/test/simple/li/li.s [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+
+  /* Other big immediates */
+  li x7, 6368
+  li x8, 0x10000042
+  li x9, 123456789
+  li x10, 0x7fffffff
+
+  /* We also allow immediates in the range [2^31, 2^32-1] (which could
+	equally have been specified as negative constants) because
+	sometimes this is more useful for "magic hex numbers" */
```
Comment (imphil) [inline hw/ip/otbn/dv/otbnsim/test/simple/li/li.s:30]: as above.

File: hw/ip/otbn/dv/otbnsim/test/simple/li/li.s [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+/* Licensed under the Apache License, Version 2.0, see LICENSE for details. */
+/* SPDX-License-Identifier: Apache-2.0 */
+/*
+    Checks for li support
+
+    This is really a test of the assembler, which is supposed to expand the li
+	pseudo-instruction into addi/lui as necessary. Of course, checking that the
+	expansion was correct is a little difficult, unless you can run the
+	generated instructions... which is what we can do for ISS testing. So the
+	checks seem to fit quite well here.
```
Comment (imphil) [inline hw/ip/otbn/dv/otbnsim/test/simple/li/li.s:11]: ALARM! TABS! :smile: ([We don't like precision.](https://www.youtube.com/watch?v=SsoOG6ZeyUI))

**Review by imphil — APPROVED** [review]

File: hw/ip/otbn/dv/otbnsim/test/simple/li/li.s (hunk 2 above)
Comment (rswarbrick) [reply hw/ip/otbn/dv/otbnsim/test/simple/li/li.s:11]: Oh no! I think I need to go and bash my `asm-mode` config...

**Pull request merged**